src/instruction.rs

```diff
@@ -0,0 +1,185 @@
+    /// amount on the system.
+    ///
+    /// NOTE: This instruction is permissionless, so anybody can finish
+    /// deactivating someone's tokens, preparing them to be withdrawn.
+    ///
+    /// 0. `[writable]` Stake config account
+    /// 1. `[writable]` Validator stake account
+    InactivateStake,
+
+    /// Withdraw inactive staked tokens from the vault
+    ///
+    /// After a deactivation has gone through the cooldown period and been
+    /// "inactivated", the authority may move the tokens out of the vault.
+    ///
+    /// 0. `[writable]` Config account
+    /// 1. `[writable]` Stake account
+    /// 2. `[writable]` Vault token account
+    /// 3. `[writable]` Destination token account
+    /// 4. `[signer]` Stake authority
+    /// 5. `[]` Vault authority, PDA with seeds `['token-owner', stake_config]`
+    /// 6. `[]` SPL Token program
+    /// 7.. Extra required accounts for transfer hook
+    ///
+    /// Instruction data: amount of tokens to move
+    WithdrawInactiveStake(u64),
+
+    /// Harvests holder SOL rewards earned by the given stake account.
+    ///
+    /// NOTE: This mostly replicates the logic in the rewards program. Since the
+    /// staked tokens are all held by this program, stakers need a way to access
+    /// their portion of holder rewards.
+    ///
+    /// This instruction requires that `unclaimed_rewards` be equal to `0` in
+    /// the token vault account. For ease of use, be sure to call the
+    /// `HarvestRewards` on the vault account before this.
+    ///
+    /// 0. `[]` Config account
+    /// 1. `[writable]` Stake account
+    /// 2. `[writable]` Vault token account
+    /// 3. `[]` Holder rewards account for vault token account
+    /// 4. `[writable]` Destination account for withdrawn lamports
+    /// 5. `[signer]` Stake authority
+    /// 6. `[]` Vault authority, PDA with seeds `['token-owner', stake_config]`
+    /// 7. `[]` Stake token mint, to get total supply
+    /// 8. `[]` SPL Token program
+    HarvestHolderRewards,
+
+    /// Harvests stake SOL rewards earned by the given stake account.
+    ///
+    /// NOTE: This is very similar to the logic in the rewards program. Since the
+    /// staking rewards are held in a separate account, they must be distributed
+    /// based on the proportion of total stake.
+    ///
+    /// 0. `[]` Config account
+    /// 1. `[writable]` Stake account
+    /// 2. `[writable]` SOL staking rewards account
+    ///    (TODO for discussion: we need a way to also track total staking rewards
+    ///    so stakers can know their allotted proportion of staking rewards,
+    ///    which are separate from the holder rewards. This means that the distribution
+    ///    logic in the Rewards program *also* needs to update some running total
+    ///    of staking rewards. I couldn't come up with a way to combine these two,
+    ///    since the proportion allocated to the different groups is *not* meant
+    ///    to be fixed forever.)
```

Comment:
Here: I think we need to have some way of tracking stake rewards. `total_rewards` at https://github.com/paladin-bladesmith/paladin-rewards-program/pull/1/files#diff-87619baf236177a7a9fc38150961151b7583b6f2f082ade75915f3ac4ca49c41R75 assumes that the distribution between holders / stakers / leader / piggybank will never change. If we want it to ever change, we need to keep the running total of holder and staker rewards separately.

And then I'm not sure where the staker rewards should be stored, but I'm open to ideas!

Reply:
this would be nice to find a solution for -- it feels like something that is likely to be up for at least a discussion of changing in the future

can we leverage a config + multi-vault architecture? where the config determines the split at a given point in time, and the vault distributes based on that config on a regular basis

Reply:
I have a few questions on this, but I'm thinking this is a good topic for tomorrow's call?

Reply:
Yeah definitely, let's cover this